core/src/providers/openai.rs

```diff
@@ -326,34 +340,48 @@ impl TryFrom<&ChatMessage> for OpenAIChatMessage {
     type Error = anyhow::Error;
 
     fn try_from(cm: &ChatMessage) -> Result<Self, Self::Error> {
-        // If `function_call_id` is present, `role` must be `function` and should be mapped to `Tool`.
-        // This is to maintain backward compatibility with the original `function` role used for content fragments.
-        let (role, tool_call_id) = match cm.function_call_id.as_ref() {
-            Some(fcid) => match OpenAIChatMessageRole::from(&cm.role) {
-                OpenAIChatMessageRole::Function => {
-                    Ok((OpenAIChatMessageRole::Tool, Some(fcid.clone())))
-                }
-                _ => Err(anyhow!(
-                    "`function_call_id` is provided but `role` is not set to `function`"
-                )),
-            },
-            _ => Ok((OpenAIChatMessageRole::from(&cm.role), None)),
-        }?;
-
-        Ok(OpenAIChatMessage {
-            content: cm.content.clone(),
-            name: cm.name.clone(),
-            role,
-            tool_call_id,
-            tool_calls: match cm.function_calls.as_ref() {
-                Some(fc) => Some(
-                    fc.into_iter()
-                        .map(|f| OpenAIToolCall::try_from(f))
-                        .collect::<Result<Vec<OpenAIToolCall>, _>>()?,
-                ),
-                None => None,
-            },
-        })
+        match cm {
+            ChatMessage::Assistant(assistant_msg) => Ok(OpenAIChatMessage {
+                content: match &assistant_msg.content {
+                    Some(c) => Some(ContentBlock::Text(c.clone())),
+                    None => None,
+                },
+                name: assistant_msg.name.clone(),
+                role: OpenAIChatMessageRole::from(&assistant_msg.role),
+                tool_calls: match assistant_msg.function_calls.as_ref() {
+                    Some(fc) => Some(
+                        fc.into_iter()
+                            .map(|f| OpenAIToolCall::try_from(f))
+                            .collect::<Result<Vec<OpenAIToolCall>, _>>()?,
+                    ),
+                    None => None,
+                },
+                tool_call_id: None,
+            }),
+            ChatMessage::Function(function_msg) => Ok(OpenAIChatMessage {
+                content: Some(ContentBlock::Text(function_msg.content.clone())),
+                name: None,
+                // If `function_call_id` is present, `role` must be `function` and should be mapped to `Tool`.
+                // This is to maintain backward compatibility with the original `function` role used for content fragments.
```

Comment:
? nothing to do with content fragment no ?

Reply:
This was already there, I just moved it around. Removed anyway since with the new types, we don't care anymore about this.